File: examples/05-production/go/concurrency_model.go
package production

import (
	"go.temporal.io/sdk/client"
	"go.temporal.io/sdk/worker"
)

// Go has no JVM virtual threads (Java's virtual_threads.java). It doesn't need
// them: each Activity invocation already runs on its own goroutine, which is
// cheap and scheduled onto a small thread pool by the runtime — the same payoff
// virtual threads give the JVM, built into the language. So worker concurrency is
// just a slot count, not a thread-pool type. Size it via worker.Options.
func StartHighConcurrencyWorker(c client.Client) worker.Worker {
	return worker.New(c, "high-concurrency-activities", worker.Options{
		// Each in-flight activity is one goroutine; this caps how many run at once.
		MaxConcurrentActivityExecutionSize: 1000,
	})
}

File: examples/05-production/python/concurrency_model.py
import concurrent.futures

from temporalio.client import Client
from temporalio.worker import Worker

# Python has no JVM virtual threads (Java's virtual_threads.java). The concurrency
# model is different: async @activity.defn functions run cooperatively on the
# worker's asyncio event loop, so thousands of I/O-bound activities can be in
# flight on a single thread without a thread-per-activity cost — the same payoff
# virtual threads give the JVM. Concurrency is capped by max_concurrent_activities.
#
# Blocking (sync def) activities, by contrast, DO need a real worker thread each;
# give the Worker an explicit thread (or process) pool executor for those.


async def start_async_worker(client: Client) -> None:
    worker = Worker(
        client,
        task_queue="high-concurrency-activities",
        workflows=[],
        activities=[],  # async @activity.defn functions: cheap, event-loop bound
        max_concurrent_activities=1000,
    )
    await worker.run()


async def start_sync_worker(client: Client) -> None:
    # Sync activities each occupy a pool thread, so size the pool deliberately.
    with concurrent.futures.ThreadPoolExecutor(max_workers=100) as pool:
        worker = Worker(
            client,
            task_queue="blocking-activities",
            workflows=[],
            activities=[],  # blocking (def) @activity.defn functions
            activity_executor=pool,
            max_concurrent_activities=100,
        )
        await worker.run()
